src/main/java/com/example/dog/service/TransactionalDogService.java

```diff
@@ -0,0 +1,57 @@
+package com.example.dog.service;
+
+import com.example.dog.dao.JdbcConnectionHolder;
+import com.example.dog.dao.ObjectNotFoundException;
+import com.example.dog.model.Dog;
+
+public class TransactionalDogService implements DogService {
+
+    private final JdbcDogService dogService;
+    private final JdbcConnectionHolder jdbcConnectionHolder;
+
+    public TransactionalDogService(JdbcDogService dogService, JdbcConnectionHolder jdbcConnectionHolder) {
+        this.dogService = dogService;
+        this.jdbcConnectionHolder = jdbcConnectionHolder;
+    }
+
+    @Override
+    public Dog get(int id) {
+        jdbcConnectionHolder.getConnection();
+        jdbcConnectionHolder.setReadOnly(true);
+        return dogService.get(id);
```

Comment:
What about committing? And handling exceptions (you should rollback).

Reply:
hmm...I didn't do it before because transaction is not started in GET method (?)

Reply:
I see. Ok then. Though not sure why you don't start tx explicitly.